### PR TITLE
Fix for #757 added channels to process one block at a time

### LINKED ISSUE #757: Event db: Ensure events processed in correct order.
[Events are emitted](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/chaincore/block/entity.go#L920)
```
go c.GetEventDb().AddEvents(b.Events)
```
and in [AddEvents](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/dbs/event/process.go#L31)
```
func (edb *EventDb) AddEvents(events []Event) {
    for loop that process events.
}
```
We need to ensure that events are processed in the correct order. That database tables are updated in the correct order.

We should protect AddEvents with a mutex or channels. So events from one block cannot be processed until events from the previous block are finished. We also do not want to slow down the chain by waiting for events to be processed. So, add an intermediate layer between the block calling AddEvents and events being processed; where execution can wait while events from earlier blocks are handled.

1. Start new event processing worker thread in [NewEventDb](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/dbs/event/eventdb.go#L8). 
2. Have the 0chain.Block thread pass the events for a block to an `event processing channel`, before continuing with the chain.
3. The event database worker thread from (1) continually polls the `event processing channel` for the next event, then processes it.

### PR DESCRIPTION
**Pull request by samtholiya**
Tasks to complete before merging PR:
- [x]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [x]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [x]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):  
...

### CODE REVIEW

**Review by kushthedude — COMMENTED** [review]
Wont the mutex block the queue for all events which are coming to be processed @samtholiya ?

**Comment by samtholiya** [comment]
> Wont the mutex block the queue for all events which are coming to be processed @samtholiya ?
image

@kushthedude 
The places where addevents is used all execute it as a goroutine. ( We can improve the code so that AddEvents itself create a goroutine)
So the parent function will not be blocked due to this...

**Comment by Sriep** [comment]
Could you check https://github.com/0chain/0chain/issues/757#issuecomment-1004040827 and see if this would work?

**Comment by kushthedude** [comment]
> Could you check [#757 (comment)](https://github.com/0chain/0chain/issues/757#issuecomment-1004040827) and see if this would work?

Mutex is better preferred to provide locking mechanism to synchronise the access to any resources, the worker thread process seems more like a semaphore & signalling mechanism which can also fix this but require more resource to keep track of each wait, signal. Why do we need to proceed with signalling mechanism @Sriep, do you see any problem here which is better solved by worker thread ?

**Comment by Sriep** [comment]
Ok maybe mutex is simpler, how would either solution work if the event database went down for a few seconds? Can mutex guarantee first in first out? https://stackoverflow.com/questions/51140111/is-sync-mutex-lock-a-fifo

**Comment by samtholiya** [comment]
@Sriep if the database goes down the channel buffer can go full so in that case even the channel acts like mutex and cannot guarantee first in first out. Refer this: https://github.com/0chain/0chain/issues/757#issuecomment-1002809781

**Comment by Sriep** [comment]
@samtholiya  

#### Channel https://github.com/0chain/0chain/issues/757#issuecomment-1004040827

If the event database is held up, then the intermediate worker thread will maintain a growing slice of events read from the ChainToWorker channel. Once the WorkerToEventD channel is no longer full, the backlog can be sent to the now with free space WorkerToEventD channel. This should ensure that events are processed in the correct order, `which is the point of` https://github.com/0chain/0chain/issues/757.

#### Mutex 

I do not see at the moment how using a mutex ensures events are processed in `first in first out` order.

#### Related changes.

New issue (coming) where we will also send an event hash for each block. Events for the latest block will need to be kept locally so they can be hashed and compared with the block event hash. As the events are coming in order it should be easy enough to compare the hash values.

**Comment by samtholiya** [comment]
@Sriep yes I do agree mutex does not handle FIFO
And I like your solution.

I would like to know what should be the breaking point.

I have once faced an issue previously where we did something similar and we found the ram consumption to be increasing exponentially in case the Database is not available for a bit longer in comparison to the number of events generated.
So there should be a breakpoint and post that something should happen... no idea what. Maybe you can help me here?

File: code/go/0chain.net/chaincore/block/entity.go [hunk 1]
```diff
@@ -995,9 +993,7 @@ func (b *Block) ComputeStateLocal(ctx context.Context, c Chainer) error {
 	}
 
 	if len(b.Events) > 0 && c.GetEventDb() != nil {
-		go func() {
-			c.GetEventDb().AddEvents(ctx, b.Events)
-		}()
+		c.GetEventDb().AddEvents(ctx, b.Events)
 		b.Events = nil
```
Comment (Sriep) [inline code/go/0chain.net/chaincore/block/entity.go:1031]: Since AddEvents is now returning an error we should log it. 
`logging.Logger.Error("..."), zap.Error(err))`

File: code/go/0chain.net/chaincore/block/entity.go [hunk 2]
```diff
@@ -905,9 +905,7 @@ func (b *Block) ComputeState(ctx context.Context, c Chainer) error {
 	}
 
 	if len(b.Events) > 0 && c.GetEventDb() != nil {
-		go func() {
-			c.GetEventDb().AddEvents(ctx, b.Events)
-		}()
+		c.GetEventDb().AddEvents(ctx, b.Events)
```
Comment (Sriep) [inline code/go/0chain.net/chaincore/block/entity.go:923]: Since AddEvents is now returning an error we should log it.

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go [hunk 3]
```diff
@@ -14,17 +20,79 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
...
+}
+
 type EventDb struct {
 	dbs.Store
+	eChannel       chan eventCtx
+	eBufferChannel chan eventCtx
+}
+
+// helps maitaining dynamic buffer for addEventWorker
+func (edb EventDb) channelBufferIntermediate() {
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: A unit test would be nice. You could generate some semi-random events. Start channelBufferIntermediate. Send your test events to eBufferChannel. Monitor eChannel and confirm it receives events in the correct order.

**Review by Sriep — COMMENTED** [review]
For manual integartion testing you could add a temporty log to ComputeStateLocal and addEventWorker. And check events are appearing correctly.

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go (hunk 3 above)
Comment (samtholiya) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: UT brought out an edge case where the goroutine is not scheduled and the channel buffer is full
in this case, the order is still not guaranteed. 
Also, why cannot we use a queue service like kafka or rabbitmq to achieve this? Yes, it will increase our requirement but if that is what we want to achieve this is the only solution.
If this was a Kubernetes-based solution. I would have recommended using new-relic to log all the events.
Please read the following golang issue where someone tried to do this
https://github.com/golang/go/issues/20352

Comment (Sriep) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: I don't follow the edge case you refer to, maybe you could commit a unit test that exposes the problem so we can see what you mean.

Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: Semaphores, mutexes, monitors, etc, in general knows nothing about data structure they serialize access to. channels is something different, it's more like data structure and semaphore together.

As I got your problem right, you can serialize access to a slice with mutext and implement FIFO logic with this slice easily, you can do queue with two slices (copy on empty) it will be more performant in some cases.
What operations you expect to prevail, reads or writes?

Comment (samtholiya) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: @dabasov no the issue is if the goroutine (that appends to the slice) does not get scheduled and the buffer for channel is full
Then the events are not serialized
cc: @Sriep

Comment (Sriep) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go]:
```go
	eventDb := &EventDb{
		Store:          db,
		eBufferChannel: make(chan eventCtx, 100),
		eChannel:       make(chan eventCtx, 100),
	}
```
Why is eBufferChannel buffered. sould it not read `eBufferChannel: make(chan eventCtx),`
```go
func (edb *EventDb) AddEvents(ctx context.Context, events []Event) {
	edb.eBufferChannel <- eventCtx{
		ctx:    ctx,
		events: events,
	}
}
```
Don't we want this to be synchronous? We are only adding the events to `buf` so this should not hold up the chain. 

Making it `eBufferChannel` buffered raises the issue if the buffer gets full. Also, we should probably change the name as we don't want it buffered.

For the moment I think we can skip the issue with running out of memory with an oversized `buf` object if the event database gets stuck. We can relook if it becomes an issue.

File: code/go/0chain.net/chaincore/block/entity.go (hunk 2 above)
Comment (dabasov) [reply code/go/0chain.net/chaincore/block/entity.go:923]: why is it done in blocking manner now without goroutine?

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go [hunk 4]
```diff
@@ -15,17 +20,81 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
...
-		Store: db,
+		Store:          db,
+		eBufferChannel: make(chan eventCtx, 100),
+		eChannel:       make(chan eventCtx, 100),
 	}
 
 	if err := eventDb.AutoMigrate(); err != nil {
 		return nil, err
 	}
+	go eventDb.channelBufferIntermediate()
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: this should be run with common.GetRootContext()

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go [hunk 5]
```diff
@@ -15,17 +20,81 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
...
+	go eventDb.addEventWorker()
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/eventdb.go:30]: this should be run with common.GetRootContext()

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go [hunk 6]
```diff
@@ -15,17 +20,81 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
...
+	buf := make([]eventCtx, 0)
+	for {
+		events, ok := <-edb.eBufferChannel
+		for _, e := range buf {
+			select {
+			case edb.eChannel <- e:
+			default:
+				break
+			}
+			buf = buf[1:]
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: I don't think that it is a good idea to drop events without any notification here, at least logging

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go [hunk 7]
```diff
@@ -15,17 +20,81 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
 		return nil, err
 	}
 	eventDb := &EventDb{
-		Store: db,
+		Store:          db,
+		eBufferChannel: make(chan eventCtx, 100),
+		eChannel:       make(chan eventCtx, 100),
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: what is an idea behind this second channel?

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go [hunk 8]
```diff
@@ -15,17 +20,81 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
...
+	}
+}
+
+// addEventWorker this worker will try to add events unless thery are not added.
+func (edb EventDb) addEventWorker() {
+	for {
+		events := <-edb.eChannel
+		for {
+			newEvents := edb.removeDuplicate(events.ctx, events.events)
+			if err := edb.addEvents(events.ctx, newEvents); err != nil && !strings.Contains(err.Error(), "len(events):0") {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/eventdb.go:50]: it might be not a good idea to pass context via event, should we write events to db if this context passed in for example computeState() got cancelled or timeout? I think yes, we better use some new context with timeout here, just to prevent long db connections

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go (hunk 7 above)
Comment (samtholiya) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go]: This was an experiment I and @Sriep were trying ... changing that.. 😓

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go (hunk 8 above)
Comment (samtholiya) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go:50]: I don't know where that event comes from ... but there are some questions
Is timeout the only context cancel reason? If yes then that can be done,
But if there are context cancellations due to other reasons how should we handle that?

Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go:50]: is there any reason for this context cancellation other than timeout?

Comment (samtholiya) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go:50]: I did not trace but we do use context.WithCancel
image
at some places 
So what if those cancellations are encountered.
Having a disconnected context will make us miss those cancellations

Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go:50]: unfortunately context is not equal to transaction. we wish to have single transactional scope for block and for event, I mean when block computation is cancelled than to cancel event writing and if block is committed (there is no such a concept yet) then commit transaction in event db. 

I agree that to ensure data consistency we have to implement something similar to distributed transaction, to be honest we don't want to use one since it will be overkill, we might:
- implement some compensation logic for cases when block state computation is cancelled, for example to generate appropriate event and to mark all generated events previously as cancelled
- we can replace event in your solution with some channel (semaphore) to start event writing after it 

the problem with your solution is that we *can* cancel the event writing if context will be cancelled before the writing, but it is not strictly

File: code/go/0chain.net/chaincore/block/entity.go (hunk 2 above)
Comment (samtholiya) [reply code/go/0chain.net/chaincore/block/entity.go:923]: errors logged. 
>why is it done in a blocking manner now without goroutine?

We are actually using the buffered channel. Which if full could block I agree but we have a buffer of 10000 as of now. Are the events going to exceed that in a short time?

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go (hunk 5 above)
Comment (samtholiya) [reply code/go/0chain.net/smartcontract/dbs/event/eventdb.go:30]: done

**Comment by samtholiya** [comment]
Old branch please refer new https://github.com/0chain/0chain/pull/968
